### ISSUE TITLE
Sketch-Guided Text-to-Image Diffusion Models

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#197]
# TL; DR

- Train a latent edge predictor that maps intermediate diffusion features to spatial edge maps, which is used to compute edge guidance that controls the generative process.

# Motivation

- Text-to-image models still lack fine-grained control. Better to incorporate simple free-hand sketches.

# Formulation

### Latent Edge Predictor

- Train a lightweight per-latent-pixel MLP network that maps intermediate diffusion features to an edge map.
- Its per-latent-pixel design makes it light and domain-agnostic.

<img width="350" alt="image" src="https://user-images.githubusercontent.com/46308407/205481403-d2336065-a386-47f4-8df6-07e90ea7ebe4.png">
<img width="360" alt="image" src="https://user-images.githubusercontent.com/46308407/205481413-cc9076a3-69a3-4ab5-b5a7-39de727b0e11.png">

- Here, P is the predictor network and F refers to channel-wise concatenation of intermediate features.
- P's input dimension is the sum of the number of channels in the used intermediate activations, and the output dimension is 4 (in the case of Stable Diffusion).

### Edge-guided Sampling

- Just like classifier-guidance (in #159 ) or reconstruction guidance ( in #183 ), compute edge reconstruction loss and use its anti-gradient for generative (reverse diffusion) process.

<img width="184" alt="image" src="https://user-images.githubusercontent.com/46308407/205481494-ab611494-37bd-4600-bd09-339120d86072.png">
<img width="354" alt="image" src="https://user-images.githubusercontent.com/46308407/205481513-f8275c04-88a0-48ac-b7dd-0305b4b0a4fc.png">

# Evaluation

- Run ImageNet images through an off-the-shelf edge detector to obtain GTs.
- It takes less than an hour on a single A100 GPU to train the latent edge predictor.

<img width="734" alt="image" src="https://user-images.githubusercontent.com/46308407/205481549-7e92ee30-6a36-491b-a945-f509b5222977.png">
<img width="358" alt="image" src="https://user-images.githubusercontent.com/46308407/205481557-d215c1ce-4cf8-4beb-ba6f-7aaddede4245.png">
<img width="354" alt="image" src="https://user-images.githubusercontent.com/46308407/205481568-1e1a4dd7-4687-42fe-b38f-c260a7ffd49f.png">
<img width="354" alt="image" src="https://user-images.githubusercontent.com/46308407/205481569-71f8ef11-f9f2-4a4c-94f6-1010971d6cb0.png">
<img width="722" alt="image" src="https://user-images.githubusercontent.com/46308407/205481565-c58b8efc-99ec-485a-a65b-fc69b3c5d817.png">
<img width="362" alt="image" src="https://user-images.githubusercontent.com/46308407/205481615-8362c005-5ceb-45b6-a546-427d874a6d64.png">
<img width="360" alt="image" src="https://user-images.githubusercontent.com/46308407/205481620-01052fea-789b-44aa-88e4-be27218a75b9.png">

# Inspiration

- The way they used a per-latent-pixel MLP to predict edges from intermediate latent features is impressive.
- I think it works because the Stable Diffusion has already mingled local information into each latent pixel, because otherwise, I think per-pixel modeling would be sup-optimal. (personal opinion)
- With the advent of text-to-image models, expanding the control boundary is becoming an extremely popular research topic. (#181 , #160 , #194 )
- I definitely have to be diligent to deliver my own work in the field, as there are just so many papers being published everyday.
